### PR TITLE
driver: crypto: hisilicon: add ECC sign and verify

### PR DESCRIPTION
**Pull request by yuzexiyzx**
add ECC sign and verify

### CODE REVIEW

**Comment by yuzexiyzx** [comment]
@etienne-lms @jforissier ECDSA is added. Do you have any comments?

File: core/drivers/crypto/hisilicon/hpre_ecc.c [hunk 1]
```diff
@@ -974,12 +1023,467 @@ static TEE_Result hpre_ecc_do_shared_secret(struct drvcrypt_secret_data *sdata)
...
+	const struct hpre_ecc_curve *curve = NULL;
+	struct ecc_public_key *pub_key = NULL;
+	TEE_Result ret = TEE_SUCCESS;
+
+	if (!sdata || !sdata->key) {
+		EMSG("Invalid ecc_verify input parameters");
+		return TEE_ERROR_BAD_PARAMETERS;
+	}
+
+	pub_key = (struct ecc_public_key *)sdata->key;
```
Comment (jenswi-linaro) [inline core/drivers/crypto/hisilicon/hpre_ecc.c:1441]: You don't need to cast a void pointer.

File: core/drivers/crypto/hisilicon/hpre_ecc.c [hunk 2]
```diff
@@ -974,12 +1023,467 @@ static TEE_Result hpre_ecc_do_shared_secret(struct drvcrypt_secret_data *sdata)
...
+	if (!pub_key) {
```
Comment (jenswi-linaro) [inline core/drivers/crypto/hisilicon/hpre_ecc.c:1442]: This can't happen, you just checked `sdata->key` above.

File: core/drivers/crypto/hisilicon/hpre_ecc.c [hunk 3]
```diff
@@ -974,12 +1023,467 @@ static TEE_Result hpre_ecc_do_shared_secret(struct drvcrypt_secret_data *sdata)
...
+	msg->param_size.ecc_sign.s_bytes = msg->curve_bytes;
+
+	return TEE_SUCCESS;
+}
+
+static TEE_Result hpre_ecdsa_param_check(struct drvcrypt_sign_data *sdata,
+					 const struct hpre_ecc_curve *curve)
+{
+	if (sdata->size_sec != BITS_TO_BYTES(curve->key_bits)) {
+		EMSG("Invalid sdata size_sec %lu", sdata->size_sec);
```
Comment (jenswi-linaro) [inline core/drivers/crypto/hisilicon/hpre_ecc.c:1162]: `%zu`

File: core/drivers/crypto/hisilicon/hpre_ecc.c [hunk 4]
```diff
@@ -479,14 +479,63 @@ static enum hisi_drv_status ecc_dh_out_to_crypto_bin(struct hpre_ecc_msg *msg)
...
+	if (ret)
+		EMSG("Fail to transfer ecc sign s from hpre_bin to crypto_bin");
+
+	return ret;
+}
+
+static enum hisi_drv_status hpre_ecc_verify_get_result(struct hpre_ecc_msg *msg,
+						       struct hpre_sqe *sqe)
+{
+	uint32_t result = (uint32_t)sqe->out;
```
Comment (jenswi-linaro) [inline core/drivers/crypto/hisilicon/hpre_ecc.c:507]: How about something like:
``` C
if (sqe->out & BIT64(1)) {
    msg->result = 1;
    return HISI_QM_DRVCRYPT_NO_ERR;
} else {
    msg->result = 0;
    return HISI_QM_DRVCRYPT_VERIFY_ERR;
}
```
The code could benefit from descriptive defines instead of the magic numbers `0` and `1`.

File: core/drivers/crypto/hisilicon/hpre_ecc.c [hunk 5]
```diff
@@ -974,12 +1023,467 @@ static TEE_Result hpre_ecc_do_shared_secret(struct drvcrypt_secret_data *sdata)
 	return ret;
 }
 
+static enum hisi_drv_status
+hpre_ecc_sign_params_fill(const struct hpre_ecc_curve *curve,
+			  struct hpre_ecc_msg *msg,
+			  struct drvcrypt_sign_data *sdata,
+			  struct bignum *rand_k)
+{
+	struct ecc_keypair *ecc_key = (struct ecc_keypair *)sdata->key;
```
Comment (jenswi-linaro) [inline core/drivers/crypto/hisilicon/hpre_ecc.c:1032]: You don't need to cast void pointers.

File: core/drivers/crypto/hisilicon/hpre_ecc.c [hunk 6]
```diff
@@ -974,12 +1023,467 @@ static TEE_Result hpre_ecc_do_shared_secret(struct drvcrypt_secret_data *sdata)
...
+	crypto_bignum_bn2bin(ecc_key->d, d);
+	ecc_sign->d_bytes = crypto_bignum_num_bytes(ecc_key->d);
+	memcpy(b, curve->b, msg->curve_bytes);
+	memcpy(n, curve->n, msg->curve_bytes);
+	memcpy(gx, curve->x, msg->curve_bytes);
+	memcpy(gy, curve->y, msg->curve_bytes);
+	crypto_bignum_bn2bin(rand_k, k);
+	ecc_sign->k_bytes = crypto_bignum_num_bytes(rand_k);
+
+	ecc_sign->e_bytes = (sdata->message.length > msg->curve_bytes ?
```
Comment (jenswi-linaro) [inline core/drivers/crypto/hisilicon/hpre_ecc.c:1056]: Please use an if statement instead of the ternary operator.
Edit: or even better:
``` C
ecc_sign->e_bytes = MIN(sdata->message.length, msg->curve_bytes);
```

File: core/drivers/crypto/hisilicon/hpre_ecc.c [hunk 7]
```diff
@@ -974,12 +1023,467 @@ static TEE_Result hpre_ecc_do_shared_secret(struct drvcrypt_secret_data *sdata)
...
+
+	return ret;
+}
+
+static enum hisi_drv_status hpre_ecc_sign_params_alloc(struct hpre_ecc_msg *msg)
+{
+	uint32_t size = HPRE_ECC_SIGN_TOTAL_BUF_SIZE(msg->key_bytes);
+	uint8_t *data = NULL;
+
+	data = (uint8_t *)calloc(1, size);
```
Comment (jenswi-linaro) [inline core/drivers/crypto/hisilicon/hpre_ecc.c:1094]: You don't need to cast void pointers.

File: core/drivers/crypto/hisilicon/hpre_ecc.c [hunk 8]
```diff
@@ -974,12 +1023,467 @@ static TEE_Result hpre_ecc_do_shared_secret(struct drvcrypt_secret_data *sdata)
...
+	msg->out = msg->in + ECC_SIGN_IN_SIZE(msg->key_bytes);
+	msg->out_dma = msg->in_dma + ECC_SIGN_IN_SIZE(msg->key_bytes);
+
+	return HISI_QM_DRVCRYPT_NO_ERR;
+}
+
+static void hpre_ecc_sign_request_deinit(struct hpre_ecc_msg *msg)
+{
+	if (msg->key) {
+		memzero_explicit(msg->key, ECC_SIGN_KEY_SIZE(msg->key_bytes));
```
Comment (jenswi-linaro) [inline core/drivers/crypto/hisilicon/hpre_ecc.c:1119]: Is this an optimization instead of using `free_wipe()`?

File: core/drivers/crypto/hisilicon/hpre_ecc.c [hunk 9]
```diff
@@ -974,12 +1023,467 @@ static TEE_Result hpre_ecc_do_shared_secret(struct drvcrypt_secret_data *sdata)
...
+	struct ecc_keypair *ecc_key = NULL;
+	struct bignum *rand_k = NULL;
+	TEE_Result ret = TEE_SUCCESS;
+
+	if (!sdata || !sdata->key) {
+		EMSG("Invalid ecc_sign input parameters");
+		return TEE_ERROR_BAD_PARAMETERS;
+	}
+
+	ecc_key = (struct ecc_keypair *)sdata->key;
```
Comment (jenswi-linaro) [inline core/drivers/crypto/hisilicon/hpre_ecc.c:1194]: You don't need to cast a void pointer.

File: core/drivers/crypto/hisilicon/hpre_ecc.c [hunk 10]
```diff
@@ -974,12 +1023,467 @@ static TEE_Result hpre_ecc_do_shared_secret(struct drvcrypt_secret_data *sdata)
...
+	if (!ecc_key) {
```
Comment (jenswi-linaro) [inline core/drivers/crypto/hisilicon/hpre_ecc.c:1195]: This can't happen, you have already checked `data->key` above.

File: core/drivers/crypto/hisilicon/hpre_ecc.c [hunk 11]
```diff
@@ -974,12 +1023,467 @@ static TEE_Result hpre_ecc_do_shared_secret(struct drvcrypt_secret_data *sdata)
...
+	memcpy(b, curve->b, msg->curve_bytes);
+	memcpy(n, curve->n, msg->curve_bytes);
+	memcpy(gx, curve->x, msg->curve_bytes);
+	memcpy(gy, curve->y, msg->curve_bytes);
+	crypto_bignum_bn2bin(ecc_key->x, pubx);
+	ecc_verify->pubx_bytes = crypto_bignum_num_bytes(ecc_key->x);
+	crypto_bignum_bn2bin(ecc_key->y, puby);
+	ecc_verify->puby_bytes = crypto_bignum_num_bytes(ecc_key->y);
+
+	ecc_verify->e_bytes = (sdata->message.length > msg->curve_bytes ?
```
Comment (jenswi-linaro) [inline core/drivers/crypto/hisilicon/hpre_ecc.c:1338]: Please use the `MIN()` macro instead.

File: core/drivers/crypto/hisilicon/hpre_ecc.c [hunk 12]
```diff
@@ -974,12 +1023,467 @@ static TEE_Result hpre_ecc_do_shared_secret(struct drvcrypt_secret_data *sdata)
...
+		return TEE_ERROR_BAD_STATE;
+
+	return TEE_SUCCESS;
+}
+
+static enum hisi_drv_status
+hpre_ecc_verify_params_alloc(struct hpre_ecc_msg *msg)
+{
+	uint32_t size = HPRE_ECC_VERIFY_TOTAL_BUF_SIZE(msg->key_bytes);
+	uint8_t *data;
```
Comment (jenswi-linaro) [inline core/drivers/crypto/hisilicon/hpre_ecc.c:1366]: This should be initialized.

File: core/drivers/crypto/hisilicon/hpre_ecc.c [hunk 13]
```diff
@@ -974,12 +1023,467 @@ static TEE_Result hpre_ecc_do_shared_secret(struct drvcrypt_secret_data *sdata)
...
+
+	data = (uint8_t *)calloc(1, size);
```
Comment (jenswi-linaro) [inline core/drivers/crypto/hisilicon/hpre_ecc.c:1368]: The cast isn't needed.

File: core/drivers/crypto/hisilicon/hpre_ecc.c [hunk 14]
```diff
@@ -974,12 +1023,467 @@ static TEE_Result hpre_ecc_do_shared_secret(struct drvcrypt_secret_data *sdata)
...
+	msg->out = msg->in + ECC_VERIF_IN_SIZE(msg->key_bytes);
+	msg->out_dma = msg->in_dma + ECC_VERIF_IN_SIZE(msg->key_bytes);
+
+	return HISI_QM_DRVCRYPT_NO_ERR;
+}
+
+static void hpre_ecc_verify_request_deinit(struct hpre_ecc_msg *msg)
+{
+	if (msg->key) {
+		memzero_explicit(msg->key, ECC_VERIF_KEY_SIZE(msg->key_bytes));
```
Comment (jenswi-linaro) [inline core/drivers/crypto/hisilicon/hpre_ecc.c:1393]: `free_wipe()` instead?

**Comment by yuzexiyzx** [comment]
@jenswi-linaro All comments have been solved.

**Comment by yuzexiyzx** [comment]
@jenswi-linaro All comments have been solved.

**Comment by yuzexiyzx** [comment]
@jenswi-linaro @jforissier All comments have been solved. Sorry for so much @, but I really need to merge all these algorithms as  soon as possible.

File: core/drivers/crypto/hisilicon/hpre_ecc.c [hunk 15]
```diff
@@ -780,8 +819,7 @@ static enum hisi_drv_status hpre_ecc_dh_params_alloc(struct hpre_ecc_msg *msg)
 static void hpre_ecc_request_deinit(struct hpre_ecc_msg *msg)
 {
 	if (msg->key) {
-		memzero_explicit(msg->key, ECC_DH_KEY_SIZE(msg->key_bytes));
-		free(msg->key);
```
Comment (jforissier) [inline core/drivers/crypto/hisilicon/hpre_ecc.c:784]: This change is OK in itself but I am wondering if there are guarantees that `free_wipe()` will always overwrite the content of the buffer. It does a regular `memset()`, it does not use the trick implemented by `memzero_explicit()` (the volatile function pointer). The generated code should be checked.
It would be good to have this change in a separate commit. It is not related to adding ECC sign and verify.

File: core/drivers/crypto/hisilicon/hpre_ecc.c [hunk 16]
```diff
@@ -510,18 +553,14 @@ static enum hisi_drv_status hpre_ecc_parse_sqe(void *bd, void *info)
 		EMSG("HPRE do ecc fail! done=0x%"PRIX16", etype=0x%"PRIX16
 		     ",etype1=0x%"PRIX16, done, err, err1);
 
-		if (done == HPRE_HW_TASK_INIT) {
-			msg->result = HISI_QM_DRVCRYPT_ENOPROC;
```
Comment (jforissier) [inline core/drivers/crypto/hisilicon/hpre_ecc.c:514]: Why setting `msg->result` is not needed anymore? It looks like an unrelated change which should go in its own commit with explanations.

File: core/drivers/crypto/hisilicon/hpre_ecc.c [hunk 17]
```diff
@@ -974,12 +1012,454 @@ static TEE_Result hpre_ecc_do_shared_secret(struct drvcrypt_secret_data *sdata)
...
+		msg->key = NULL;
+	}
+}
+
+static TEE_Result
+hpre_ecc_verify_request_init(const struct hpre_ecc_curve *curve,
+			     struct hpre_ecc_msg *msg,
+			     struct drvcrypt_sign_data *sdata)
+{
+	int32_t ret;
```
Comment (jforissier) [inline core/drivers/crypto/hisilicon/hpre_ecc.c:1385]: = 0;

File: core/drivers/crypto/hisilicon/hpre_ecc.c (hunk 16 above)
Comment (yuzexiyzx) [reply core/drivers/crypto/hisilicon/hpre_ecc.c:514]: we judge return value instead of msg->result, so msg->result is useless

File: core/drivers/crypto/hisilicon/hpre_ecc.c (hunk 15 above)
Comment (yuzexiyzx) [reply core/drivers/crypto/hisilicon/hpre_ecc.c:784]: OK, I will fix this after this commit done.

File: core/drivers/crypto/hisilicon/hpre_ecc.c (hunk 16 above)
Comment (yuzexiyzx) [reply core/drivers/crypto/hisilicon/hpre_ecc.c:514]: maybe I should use another pr to fix these two problem? These are not related to ecc sign and verify indeed.

File: core/drivers/crypto/hisilicon/hpre_ecc.c (hunk 17 above)
Comment (yuzexiyzx) [reply core/drivers/crypto/hisilicon/hpre_ecc.c]: solved

File: core/drivers/crypto/hisilicon/hpre_ecc.c (hunk 16 above)
Comment (yuzexiyzx) [reply core/drivers/crypto/hisilicon/hpre_ecc.c:514]: solved

File: core/drivers/crypto/hisilicon/hpre_ecc.c (hunk 15 above)
Comment (yuzexiyzx) [reply core/drivers/crypto/hisilicon/hpre_ecc.c:784]: solved

**Comment by yuzexiyzx** [comment]
@jforissier All comments have been solved. This failing may be not related to my modification. I can not find error about ecc.

**Review by jforissier — COMMENTED** [review]
- Please use parentheses with function names like so: "driver: crypto: hisilicon: replace free() and memzero() by free_wipe()"

  "Replace free() and memzero() by free_wipe()."

For all commits:

`Acked-by: Jerome Forissier <jerome.forissier@linaro.org>`

**Comment by yuzexiyzx** [comment]
@jforissier All comments have been solved

**Comment by yuzexiyzx** [comment]
@jforissier Can this PR be merged?

**Comment by jforissier** [comment]
> @jforissier Can this PR be merged?

Yes, thanks.
